ding/model/template/vae.py

```diff
@@ -0,0 +1,210 @@
+        raise RuntimeWarning()
+
+    def generate(self, x: Tensor, **kwargs) -> Tensor:
+        raise NotImplementedError
+
+    @abstractmethod
+    def forward(self, *inputs: Tensor) -> Tensor:
+        pass
+
+    @abstractmethod
+    def loss_function(self, *inputs: Any, **kwargs) -> Tensor:
+        pass
+
+
+class VanillaVAE(BaseVAE):
+
+    def __init__(self, action_dim: int, obs_dim: int, latent_dim: int, hidden_dims: List = None, **kwargs) -> None:
+        super(VanillaVAE, self).__init__()
+
+        self.action_dim = action_dim
+        self.obs_dim = obs_dim
+        self.latent_dim = latent_dim
+        self.hidden_dims = hidden_dims
+
+        modules = []
+        if hidden_dims is None:
+            hidden_dims = [256]
+
+        # Build Encoder
+        # action
+        self.encode_action_head = nn.Sequential(nn.Linear(self.action_dim, hidden_dims[0]), nn.ReLU())
+        # obs
+        self.encode_obs_head = nn.Sequential(nn.Linear(self.obs_dim, hidden_dims[0]), nn.ReLU())
+
+        self.encode_common = nn.Sequential(nn.Linear(hidden_dims[0], hidden_dims[0]), nn.ReLU())
+        self.encode_mu_head = nn.Linear(hidden_dims[0], latent_dim)
+        self.encode_logvar_head = nn.Linear(hidden_dims[0], latent_dim)
+
+        # Build Decoder
+        self.condition_obs = nn.Sequential(nn.Linear(self.obs_dim, hidden_dims[0]), nn.ReLU())
+        self.decode_action_head = nn.Sequential(nn.Linear(latent_dim, hidden_dims[0]), nn.ReLU())
+        self.decode_common = nn.Sequential(nn.Linear(hidden_dims[0], hidden_dims[0]), nn.ReLU())
+        # TODO(pu): tanh
+        self.decode_reconst_action_head = nn.Sequential(nn.Linear(hidden_dims[0], self.action_dim), nn.Tanh())
+        # self.decode_reconst_action_head = nn.Linear(hidden_dims[0], self.action_dim)
+
+        # residual prediction
+        self.decode_prediction_head_layer1 = nn.Sequential(nn.Linear(hidden_dims[0], hidden_dims[0]), nn.ReLU())
+        self.decode_prediction_head_layer2 = nn.Linear(hidden_dims[0], self.obs_dim)
+
+        self.obs_encoding = None
+
+    def encode(self, input) -> List[Tensor]:
+        """
+        Encodes the input by passing through the encoder network
+        and returns the latent codes.
+        :param input: (Tensor) Input tensor to encoder [N x C x H x W]
+        :return: (Tensor) List of latent codes
+        """
+        action_encoding = self.encode_action_head(input['action'])
+        obs_encoding = self.encode_obs_head(input['obs'])
+        # obs_encoding = self.condition_obs(input['obs'])  #  TODO(pu): using a different network
+
+        self.obs_encoding = obs_encoding
+        # input = torch.cat([obs_encoding, action_encoding], dim=-1)
+        # input = obs_encoding + action_encoding  # TODO(pu): what about add, cat?
+        input = obs_encoding * action_encoding
+
+        result = self.encode_common(input)
+        result = torch.flatten(result, start_dim=1)
```

Comment:
use our Flatten for compatibility of different torch version (`from ding.torch_utils import Flatten`)